core/src/main/java/io/grpc/Context.java

```diff
@@ -591,17 +621,18 @@ public boolean isCurrent() {
 
     /**
      * Attach this context to the thread and return a {@link AutoCloseable} that can be
-     * used with try-with-resource statements to properly {@link #detach} and {@link #cancel}
-     * the context on completion.
+     * used with try-with-resource statements to properly attach the previously bound context
+     * when {@link AutoCloseable#close()} is called.
      *
      * @return a {@link java.io.Closeable} which can be used with try-with-resource blocks.
      */
     public Closeable attachAsCloseable() {
-      attach();
+      final Context previous = attach();
       return new Closeable() {
         @Override
         public void close() throws IOException {
-          detachAndCancel(null);
+          cancel(null);
+          previous.attach();
```

Comment:
Do we want to encourage attaching the previous instead of calling detach? If not, then `detach(previous)` would be best here.